pkg/kube/inject/webhook.go

```diff
@@ -568,23 +571,53 @@ func reapplyOverwrittenContainers(finalPod *corev1.Pod, originalPod *corev1.Pod,
 		finalPod.Annotations[annotation.ProxyOverrides.Name] = string(js)
 	}
 
+	adjustInitContainerUser(finalPod, originalPod)
+
 	return finalPod, nil
 }
 
-func resetFieldsInAutoImageContainer(original *corev1.Container, template *corev1.Container) {
-	if original.Image == AutoImage {
-		original.Image = ""
+// adjustInitContainerUser adjusts the RunAsUser/Group fields and iptables parameter "-u <uid>"
+// in the init/validation container so that they match the value of SecurityContext.RunAsUser/Group
+// when it is present in the custom istio-proxy container supplied by the user.
+func adjustInitContainerUser(finalPod *corev1.Pod, originalPod *corev1.Pod) {
+	userContainer := FindSidecar(originalPod)
+	if userContainer == nil {
+		// if user doesn't override the istio-proxy container, there's nothing to do
+		return
 	}
 
-	// If the original pod comes with SecurityContext.RunAsUser and the template defines a value different than the default (1337),
-	// then ignore the original value and stick with the final (merged one)
-	// This is likely a scenario in OpenShift when the istio-proxy container with image: auto is parsed, if SecurityContext.RunAsUser
-	// does not exist, OpenShift automatically assigns a value which is based on an annotation in the namespace. Regardless if the user
-	// provided that value or if it was assigned by OpenShift, the correct value is the one in the template, as set by the `.ProxyUID` field.
-	if original.SecurityContext != nil && template.SecurityContext != nil && template.SecurityContext.RunAsUser != nil &&
-		*template.SecurityContext.RunAsUser != iptablesConstants.DefaultProxyUIDInt {
-		original.SecurityContext.RunAsUser = nil
-		original.SecurityContext.RunAsGroup = nil
+	if userContainer.SecurityContext == nil || (userContainer.SecurityContext.RunAsUser == nil && userContainer.SecurityContext.RunAsGroup == nil) {
+		// if user doesn't override SecurityContext.RunAsUser/Group, there's nothing to do
+		return
+	}
+
+	// Locate the istio-init or istio-validation container
+	initContainer := FindContainer(InitContainerName, finalPod.Spec.InitContainers)
+	if initContainer == nil {
+		initContainer = FindContainer(ValidationContainerName, finalPod.Spec.InitContainers)
+	}
+	if initContainer == nil {
+		// should not happen
+		return
+	}
```

Comment:
Wdyt about the following for readability? It's not immediately obvious why we are checking initContainer for nil twice and calling FindContainer twice until you really look at the input values and see that there are two different names. Something like the code suggestion below might make that fact more obvious:

```suggestion
        // Locate the istio-init or istio-validation container
        var initContainer *Container
        for _, name := range []string{InitContainerName, ValidationContainerName} {
            if container := FindContainer(name, finalPod.Spec.InitContainers); container != nil {
                initContainer = container
            }
        }
        if initContainer == nil {
            // should not happen
            return
        }
```